### PR TITLE
Use RCP event bus instead of the Salt notification extension

### PR DESCRIPTION
**Pull request by thomaskrause**
## Pull request type

 

Feature and Refactoring (no functional changes, no API changes)

## What is the current behavior?

We are relying on the Salt notification mechanism to handle automatic redraws of the views when the Salt graph was changed and other things like the dirty state indicator.
This Salt extension is quite buggy and already caused several bugs. It also has some problematic design choices in regard to which objects are given as argument when a change event is reported.
When trying to implement undo/redo functionality, it became obvious, that  the current notification extension of Salt is not sufficient.

Issue Number: N/A

## What is the new behavior?

- We provide our own Salt factory and Salt Graph objects which send events using the Eclipse event bus
- Events are sent after a Salt element was added, before it is removed and before and after a modification was performed
- The element that was changed is given as argument. This is different from the Salt notifications, where the change type (add, delete, set, ...) and some other information like the old and new value are given. Salt notifications also used the delegated objects which made it impossible to check the Salt type of the argument. This implemention will provide the actual Salt Common object (like `SDocumentGraph`). Since we get an event before a change is happening, we could save the relevant information (like all attributes of a node) before they are actually changed. Also, using Eclipse event topics allow us to introduce sub-topics (like `ANNOTATION_UPDATE/ADD`/`ANNOTATION_UPDATE/REMOVED` etc.) with more fine-grained semantics. These semantics might be extended or refactored when undo/redo functionality is implemented.

## Does this introduce a breaking change?

The `addListener`/`removeListener` functions from the `ProjectManager` have been removed because you can now directly attach yourself to the event topics using RCP.

## Pull request checklist

Please check if your PR fulfills the following requirements:
- [X] Tests for the changes have been added **or** the PR is neither a bug fix nor a feature
- [X] Docs have been reviewed and added / updated if needed **or** the PR is neither a bug fix nor a feature
- [X] The *Unreleased* section in CHANGELOG.md has been amended to reflect the changes in this PR if needed
- [X] Build (`mvn verify`) was run locally and any changes were pushed
- [X] The pull request is against the correct branch (`master` for bug fixes, `develop` for new functionality)
- [X] [Dependencies and citation templates](https://github.com/hexatomic/hexatomic/tree/develop/releng/templates) have been updated where necessary **or** there are no new dependencies

---

# Checklist for maintainers

## Releases

- Do **NOT** push the green merge button on GitHub.  
Follow the latest version of the [developer/maintainer docs](https://hexatomic.github.io/hexatomic/dev/) for making releases.
- Check that license and citation information for dependencies are complete (each dependency has a folder in `/THIRD-PARTY/`).

### CODE REVIEW

**Comment by thomaskrause** [comment]
SonarCloud actually found some issues I should address, making this a draft again.

**Comment by thomaskrause** [comment]
Re-opening this PR after fixing some things reported by the static code analysis.
Please note that the reported test coverage is wrong (would be fixed by #189) and should be ignored.
The duplicated lines in the various notification graph element classes are a necessary evil, since this can't be moved into a common parent class.

**Comment by thomaskrause** [comment]
> Re-opening this PR after fixing some things reported by the static code analysis.
> Please note that the reported test coverage is wrong (would be fixed by #189) and should be ignored.
> The duplicated lines in the various notification graph element classes are a necessary evil, since this can't be moved into a common parent class.

I merged develop into the PR to get the more realistic code coverage, but now it also includes the test bundles in the code coverage which drags the metric down. I try to set a corresponding property in the SonarCloud administration interface to exclude our tests bundles from the code coverage report. I think it is not recognized automatically as tests because they are not in the src/test folder but in src/main

**Comment by thomaskrause** [comment]
> > Re-opening this PR after fixing some things reported by the static code analysis.
> > Please note that the reported test coverage is wrong (would be fixed by #189) and should be ignored.
> > The duplicated lines in the various notification graph element classes are a necessary evil, since this can't be moved into a common parent class.
> 
> I merged develop into the PR to get the more realistic code coverage, but now it also includes the test bundles in the code coverage which drags the metric down. I try to set a corresponding property in the SonarCloud administration interface to exclude our tests bundles from the code coverage report. I think it is not recognized automatically as tests because they are not in the src/test folder but in src/main

The code coverage is now correct, the duplicated lines are still necessary (and I reported this a false positive to SonarCloud).

**Comment by thomaskrause** [comment]
I updated the PR to include less code duplication (and to make it less probable that something is changed only in one spot) but even with the changes, some amount of duplication is necessary. I therefore excluded the notification graph element java files from the duplication analysis in SonarCloud

File: bundles/org.corpus_tools.hexatomic.core/src/main/java/org/corpus_tools/hexatomic/core/events/salt/SaltNotificationFactory.java [hunk 1]
```diff
@@ -0,0 +1,529 @@
...
+   * @throws IllegalStateException Thrown if the current {@link SaltFactory} implementation is not
+   *         of this type.
+   */
+  static void sendEvent(String topic, Object element) {
+
+    // Get an instance to the factory singleton and check if this has the correct class
+    ISaltFactory factory = SaltFactory.getFactory();
+    if (factory instanceof SaltNotificationFactory) {
+      SaltNotificationFactory notificationFactory = (SaltNotificationFactory) factory;
+      // We have to assume that other thread can modify the Salt graph.
```
Comment (sdruskat) [inline bundles/org.corpus_tools.hexatomic.core/src/main/java/org/corpus_tools/hexatomic/core/events/salt/SaltNotificationFactory.java:187]:
```suggestion
      // We have to assume that other threads can modify the Salt graph.
```

Fix typo.

File: bundles/org.corpus_tools.hexatomic.core/src/main/java/org/corpus_tools/hexatomic/core/SaltHelper.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+ * @author Thomas Krause {@literal <krauseto@hu-berlin.de>}
+ */
+public class SaltHelper {
+  
+  private SaltHelper() {
+    // Should not be initiated
+  }
+
+  /**
+   * Delegated elements might have an actual owning element. This functions returns the actual
```
[the remark this answers is not in the document]
Comment (thomaskrause) [reply bundles/org.corpus_tools.hexatomic.core/src/main/java/org/corpus_tools/hexatomic/core/SaltHelper.java]: Updated in f9d5d50f0ae0644dd8bd5bdefbb27ba22c3086ab

**Pull request merged**